lib/IMAP/MessageMapper.php

```diff
@@ -161,30 +161,31 @@ public function findAll(Horde_Imap_Client_Socket $client,
 			$this->logger->debug("Range for findAll did not find any messages. Trying again with a succeeding range");
 			return $this->findAll($client, $mailbox, $maxResults, $upper);
 		}
+
+		$uidCandidates = array_map(
+			function (Horde_Imap_Client_Data_Fetch $data) {
+				return $data->getUid();
+			},
+			iterator_to_array($fetchResult)
+		);
+		// We don't know if the server already sorted the UIDs
+		sort($uidCandidates);
+
 		$uidsToFetch = array_slice(
-			array_filter(
-				array_map(
-					function (Horde_Imap_Client_Data_Fetch $data) {
-						return $data->getUid();
-					},
-					iterator_to_array($fetchResult)
-				),
-
-				function (int $uid) use ($highestKnownUid) {
-					// Don't load the ones we already know
-					return $uid > $highestKnownUid;
-				}
-			),
+			$uidCandidates,
 			0,
 			$maxResults
 		);
+
+		$highestUidFetched = $uidsToFetch[count($uidsToFetch) - 1];
+		$this->logger->debug(sprintf("Range for findAll min=$min max=$max found %d messages, %d left after filtering. Highest UID fetched is %d", count($uidCandidates), count($uidsToFetch), $highestUidFetched));
 		return [
 			'messages' => $this->findByIds(
 				$client,
 				$mailbox,
 				$uidsToFetch
 			),
-			'all' => $upper === $max,
+			'all' => $highestUidFetched === $max,
```

Comment:
this is the important fix. expected highest UID ($upper) vs actual highest fetched UID ($highestUidFetched)